File: 2022-05/python/203_Remove_Linked List_Elements.py
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        return self.recursively(head, val)
        return self.iteratively(head, val)

    def recursively(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        if not head: return None
        head.next = self.recursively(head.next, val)
        return head.next if head.val == val else head

    def iteratively(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        start = ListNode(-1, head)
        cur = start
        while cur.next:
            if cur.next.val == val:
                cur.next = cur.next.next
            else:
                cur = cur.next
        return start.next

File: 2022-05/javascript/203_Remove_Linked List_Elements.js

function removeElements(head, val) {
    return recursively(head, val)
    // return iteratively(head, val)
}

function recursively(head, val) {
    if (!head) return null
    head.next = recursively(head.next, val)
    return head.val === val ? head.next : head
}

function iteratively(head, val) {
    const start = new ListNode(-1, head)
    let cur = start
    while (cur.next) {
        if (cur.next.val === val) {
            cur.next = cur.next.next
        } else {
            cur = cur.next
        }
    }
    return start.next
}